s3transfer/download.py

```diff
@@ -0,0 +1,294 @@
+            placed in
+        :param start_index: The location in the file to start writing the
+            content of the key to.
+        """
+        last_exception = None
+        for i in range(max_attempts):
+            try:
+                response = client.get_object(
+                    Bucket=bucket, Key=key, **extra_args)
+                streaming_body = StreamReaderProgress(
+                    response['Body'], callbacks)
+
+                current_index = start_index
+                chunks = iter(
+                    lambda: streaming_body.read(self.STREAM_CHUNK_SIZE), b'')
+                for chunk in chunks:
+                    # If the io queue was shutdown just stop sending to
+                    # chunks to the queue and quit the download.
+                    try:
+                        io_queue.put((current_index, chunk))
+                    except QueueShutdownError:
+                        return
+                    current_index += len(chunk)
+                return
+            except S3_RETRYABLE_ERRORS as e:
+                logger.debug("Retrying exception caught (%s), "
+                             "retrying request, (attempt %s / %s)", e, i,
+                             max_attempts, exc_info=True)
+                last_exception = e
+                continue
+        raise RetriesExceededError(last_exception)
+
+
+class IOWriteTask(Task):
+    def _main(self, filename, final_filename, osutil, io_queue):
+        """Pulls off an io queue to write contents to a file
+
+        :param filename: The name of the file to write contents to
+        :param final_filename: The final name of the file to rename to
+            upon completion of writing the contents.
+        :param osutil: OS utility
+        :param io_queue: The queue to retrieve contents of the downloaded
+            file off of.
+        """
+        self._transfer_coordinator.add_failure_cleanup(
+            self._remove_file_cleanup, filename, osutil)
+        try:
+            self._loop_on_io_writes(filename, osutil, io_queue)
+        except Exception as e:
+            logger.debug("Caught exception in IO thread: %s",
+                         e, exc_info=True)
+            io_queue.trigger_shutdown()
+            raise
+        if self._transfer_coordinator.status != 'failed':
+            osutil.rename_file(filename, final_filename)
+
+    def _loop_on_io_writes(self, filename, osutil, io_queue):
+        with osutil.open(filename, 'wb') as f:
+            while True:
+                task = io_queue.get()
+                if task is SHUTDOWN_SENTINEL:
+                    logger.debug("Shutdown sentinel received in IO handler, "
+                                 "shutting down IO handler.")
+                    return
+                else:
+                    offset, data = task
+                    f.seek(offset)
+                    f.write(data)
+
+    def _remove_file_cleanup(self, filename, osutil):
```

Comment:
I wonder if that is a bug. I will look into it. This is where it is suppose to be cleaned up: https://github.com/boto/s3transfer/pull/5/files#diff-a12a8d2ff65118bd0968bf807b8cf8cdR261. Adding it as a failure_cleanup through the transfer context should handle it.